File: sdks/shrinkix-php/src/Transport.php
<?php

namespace Shrinkix;

use GuzzleHttp\Client as GuzzleClient;
use GuzzleHttp\Exception\GuzzleException;
use Shrinkix\Errors\ApiException;
use Shrinkix\Errors\NetworkException;

/**
 * HTTP Transport Layer
 */
class Transport
{
    private string $apiKey;
    private string $baseUrl;
    private bool $sandbox;
    private GuzzleClient $client;

    public function __construct(string $apiKey, string $baseUrl = 'https://api.shrinkix.com/v1', bool $sandbox = false)
    {
        $this->apiKey = $apiKey;
        $this->baseUrl = $baseUrl;
        $this->sandbox = $sandbox;

        $headers = [
            'Authorization' => "Bearer {$apiKey}",
            'User-Agent' => 'shrinkix-php/1.0.0',
        ];

        if ($sandbox) {
            $headers['X-Mode'] = 'sandbox';
        }

        $this->client = new GuzzleClient([
            'base_uri' => $baseUrl,
            'headers' => $headers,
        ]);
    }

    public function get(string $endpoint, array $params = []): array
    {
        return $this->request('GET', $endpoint, ['query' => $params]);
    }

    public function post(string $endpoint, array $options = []): array
    {
        return $this->request('POST', $endpoint, $options);
    }

    private function request(string $method, string $endpoint, array $options = []): array
    {
        try {
            $response = $this->client->request($method, $endpoint, $options);

            // Extract rate limit headers
            $rateLimit = [
                'limit' => $response->getHeaderLine('x-ratelimit-limit'),
                'remaining' => $response->getHeaderLine('x-ratelimit-remaining'),
                'reset' => $response->getHeaderLine('x-ratelimit-reset'),
                'request_id' => $response->getHeaderLine('x-request-id'),
            ];

            $body = json_decode($response->getBody()->getContents(), true);

            return [
                'data' => $body,
                'rate_limit' => $rateLimit,
                'headers' => $response->getHeaders(),
            ];

        } catch (GuzzleException $e) {
            if ($e->hasResponse()) {
                $response = $e->getResponse();
                $body = json_decode($response->getBody()->getContents(), true);

                $rateLimit = [
                    'limit' => $response->getHeaderLine('x-ratelimit-limit'),
                    'remaining' => $response->getHeaderLine('x-ratelimit-remaining'),
                    'reset' => $response->getHeaderLine('x-ratelimit-reset'),
                    'request_id' => $response->getHeaderLine('x-request-id'),
                ];

                throw new ApiException(
                    $body['message'] ?? 'API Error',
                    $body['error'] ?? 'UNKNOWN_ERROR',
                    $response->getStatusCode(),
                    $body['request_id'] ?? null,
                    $body['details'] ?? [],
                    $body['docs_url'] ?? null,
                    $rateLimit,
                    $response->getHeaderLine('retry-after') ?: null
                );
            }

            throw new NetworkException('Network request failed', $e);
        }
    }
}

File: sdks/shrinkix-ruby/lib/shrinkix/transport.rb
# frozen_string_literal: true

require 'faraday'
require 'faraday/multipart'
require_relative 'errors'

module Shrinkix
  # HTTP Transport Layer
  class Transport
    def initialize(api_key:, base_url: 'https://api.shrinkix.com/v1', sandbox: false)
      @api_key = api_key
      @base_url = base_url
      @sandbox = sandbox
      @connection = build_connection
    end

    def get(endpoint, params: {})
      request(:get, endpoint, params: params)
    end

    def post(endpoint, body: nil, multipart: false)
      request(:post, endpoint, body: body, multipart: multipart)
    end

    private

    def build_connection
      Faraday.new(url: @base_url) do |conn|
        conn.request :multipart if @sandbox
        conn.request :url_encoded
        conn.response :json, content_type: /\bjson$/
        conn.adapter Faraday.default_adapter

        conn.headers['Authorization'] = "Bearer #{@api_key}"
        conn.headers['User-Agent'] = 'shrinkix-ruby/1.0.0'
        conn.headers['X-Mode'] = 'sandbox' if @sandbox
      end
    end

    def request(method, endpoint, params: {}, body: nil, multipart: false)
      response = @connection.public_send(method) do |req|
        req.url endpoint
        req.params = params if params.any?
        req.body = body if body

        if multipart
          req.headers['Content-Type'] = 'multipart/form-data'
        end
      end

      # Extract rate limit headers
      rate_limit = {
        limit: response.headers['x-ratelimit-limit'],
        remaining: response.headers['x-ratelimit-remaining'],
        reset: response.headers['x-ratelimit-reset'],
        request_id: response.headers['x-request-id']
      }

      # Handle errors
      unless response.success?
        body = response.body
        raise ApiError.new(
          body['message'] || 'API Error',
          code: body['error'] || 'UNKNOWN_ERROR',
          status_code: response.status,
          request_id: body['request_id'],
          details: body['details'] || {},
          docs_url: body['docs_url'],
          rate_limit: rate_limit,
          retry_after: response.headers['retry-after']
        )
      end

      {
        data: response.body,
        rate_limit: rate_limit,
        headers: response.headers
      }
    rescue Faraday::Error => e
      raise NetworkError.new('Network request failed', original_error: e)
    end
  end
end
